secboot/keymgr/keymgr_luks2.go

```diff
@@ -0,0 +1,229 @@
+	keyringAddKeyToUserKeyring   = keyring.AddKeyToUserKeyring
+)
+
+func getEncryptionKeyFromUserKeyring(dev string) ([]byte, error) {
+	const remove = false
+	const defaultPrefix = "ubuntu-fde"
+	// note this is the unlock key, which can be either the main key which
+	// was unsealed, or the recovery key, in which case some operations may
+	// not make sense
+	currKey, err := sbGetDiskUnlockKeyFromKernel(defaultPrefix, dev, remove)
+	if err != nil {
+		return nil, fmt.Errorf("cannot obtain current unlock key for %v: %v", dev, err)
+	}
+	return currKey, err
+}
+
+func isKeyslotNotActive(err error) bool {
+	match, _ := regexp.MatchString(`.*: Keyslot [0-9]+ is not active`, err.Error())
+	return match
+}
+
+func recoveryKDF() (*sb.KDFOptions, error) {
+	usableMem, err := osutil.TotalUsableMemory()
+	if err != nil {
+		return nil, fmt.Errorf("cannot get usable memory for KDF parameters when adding the recovery key: %v", err)
+	}
+	// The KDF memory is heuristically calculated by taking the
+	// usable memory and subtracting hardcoded 384MB that is
+	// needed to keep the system working. Half of that is the mem
+	// we want to use for the KDF. Doing it this way avoids the expensive
+	// benchmark from cryptsetup. The recovery key is already 128bit
+	// strong so we don't need to be super precise here.
+	kdfMem := (int(usableMem) - 384*1024*1024) / 2
+	// max 1 GB
+	if kdfMem > 1024*1024*1024 {
+		kdfMem = (1024 * 1024 * 1024)
+	}
+	// min 32 KB
+	if kdfMem < 32*1024 {
+		kdfMem = 32 * 1024
+	}
+	return &sb.KDFOptions{
+		MemoryKiB:       kdfMem / 1024,
+		ForceIterations: 4,
+	}, nil
+}
+
+func sbKDFToLuksKDF(o *sb.KDFOptions) luks2.KDFOptions {
+	return luks2.KDFOptions{
+		TargetDuration:  o.TargetDuration,
+		MemoryKiB:       o.MemoryKiB,
+		ForceIterations: o.ForceIterations,
+		Parallel:        o.Parallel,
+	}
+}
+
+// AddRecoveryKeyToLUKSDevice adds a recovery key to a LUKS2 device. It the
+// devuce unlock key from the user keyring to authorize the change. The
+// recoveyry key is added to keyslot 1.
+func AddRecoveryKeyToLUKSDevice(dev string, recoveryKey secboot.RecoveryKey) error {
+	currKey, err := getEncryptionKeyFromUserKeyring(dev)
+	if err != nil {
+		return err
+	}
+
+	return AddRecoveryKeyToLUKSDeviceUsingKey(dev, recoveryKey, currKey)
+}
+
+// AddRecoveryKeyToLUKSDeviceUsingKey adds a recovery key rkey to the existing
+// LUKS encrypted volume on the block device given by node. The existing key to
+// the encrypted volume is provided in the key argument and used to authorize
+// the operation.
+//
+// A heuristic memory cost is used.
+func AddRecoveryKeyToLUKSDeviceUsingKey(dev string, recoveryKey secboot.RecoveryKey, currKey secboot.EncryptionKey) error {
+	opts, err := recoveryKDF()
+	if err != nil {
+		return err
+	}
+
+	if err := luks2.KillSlot(dev, recoveryKeySlot, currKey[:]); err != nil {
+		if !isKeyslotNotActive(err) {
+			return fmt.Errorf("cannot kill existing slot: %v", err)
+		}
+	}
+	// TODO: fixup options?
```

Comment:
what's the issue here?

Reply:
Just a leftover thing I missed. It did some KDF options manipulation before.